nvme.c

```diff
@@ -288,47 +293,67 @@ static int get_telemetry_log(int argc, char **argv, struct command *cmd, struct
 		goto close_output;
 	}
 
-	err = write(output, (void *) hdr, 4096);
-	if (err != 4096) {
+	err = write(output, (void *) hdr, bs);
+	if (err != bs) {
 		fprintf(stderr, "Failed to flush all data to file!");
 		goto free_hdr;
 	}
-
-	num_blocks = max(hdr->dalb1, max(hdr->dalb2, hdr->dalb3));
-
-	full_size = num_blocks * 512;
-	/* Round to page boundary */
-	full_size += (4096 - (full_size % 4096));
-	/* We've already pulled 1 page worth of data, lets remove that. */
-	full_size -= 4096;
-	offset = 4096;
-
-	page_log = malloc(4096);
+	/* Setting offset at blocksize since the first block is the header
+	 * and is already pulled
+	 */
+	offset = bs;
+	page_log = malloc(bs);
 	if (!page_log) {
 		fprintf(stderr, "Failed to allocate %zu bytes for log\n",
-			full_size);
+			bs);
 		err = ENOMEM;
 		goto free_hdr;
 	}
 
-	while (full_size) {
-		err = nvme_get_telemetry_log(fd, page_log, 0, cfg.ctrl_init, 4096, offset);
+	/* DALB3 should always be the largest value according to the spec
+	 * Since grabbing log page 3 and grabbing all telemetry log data
+	 * is functionally identical, we can toss it in the same condition
+	 * with the or binary operator.
+	 */
+	if (!cfg.data_area || cfg.data_area == 3) {
+		full_size = (hdr->dalb3 * bs) + offset;
+	} else if (cfg.data_area == 1) {
+		full_size = (hdr->dalb1 * bs) + offset;
+	} else if (cfg.data_area == 2) {
+		full_size = (hdr->dalb2 * bs) + offset;
+	} else {
+		fprintf(stderr, "Invalid data area requested");
+		goto free_hdr;
```

Comment:
page_log leaks here if error is occurred..

Reply:
err = EINVAL in this default case.